source/agora/consensus/Quorum.d

```diff
@@ -0,0 +1,1229 @@
+        {
+            return *name;
+        }
+        else
+        {
+            string name = names[last_used];
+            last_used++;
+
+            if (last_used >= names.length)
+                assert(0);  // add more names plz
+
+            hashToName[hash] = name;
+            return name;
+        }
+    }
+
+    return toUserReadable(input[].hashFull());
+}
+
+// wrapper around Amount with some unsafe primitives
+// note: temporary, will remove soon
+private struct TestAmount
+{
+    Amount amount;
+
+    TestAmount opBinary (string op : "+", T)(in T rhs)
+    {
+        assert(this.amount.add(*cast(Amount*)&rhs));
+        return this;
+    }
+
+    TestAmount opBinary (string op : "-", T)(in T rhs)
+    {
+        assert(this.amount.sub(*cast(Amount*)&rhs));
+        return this;
+    }
+
+    TestAmount opBinary (string op : "*")(size_t count)
+    {
+        if (count == 0)
+        {
+            this.amount = Amount.init;
+        }
+        else
+        {
+            foreach (_; 0 .. count - 1)
+                this.amount.add(this.amount);
+        }
+
+        return this;
+    }
+}
+
+/// Generate a tuple pair of (Enrollment[], UTXOFinder)
+version (unittest)
+private auto genEnrollsAndFinder (size_t enroll_count,
+    TestAmount delegate (TestAmount) getAmount,
+    TestAmount initial_amount = TestAmount(Amount.MinFreezeAmount))
+{
+    import agora.common.Amount;
+    import agora.consensus.data.Transaction;
+    import agora.consensus.Genesis;
+
+    TestUTXOSet storage = new TestUTXOSet;
+    Enrollment[] enrolls;
+
+    TestAmount prev_amount;
+    foreach (idx; 0 .. enroll_count)
+    {
+        TestAmount amount;
+        if (idx == 0)
+            amount = initial_amount;
+        else
+            amount = getAmount(prev_amount);
+        Transaction tx =
+        {
+            type : TxType.Freeze,
+            outputs: [Output(amount.amount, getTestKey(idx))]
+        };
+
+        storage.put(tx);
+        prev_amount = amount;
+    }
+
+    foreach (utxo; storage.keys)
+    {
+        Enrollment enroll =
+        {
+            utxo_key : utxo,
+            cycle_length : 1008
+        };
+
+        enrolls ~= enroll;
+    }
+
+    return tuple(enrolls, &storage.findUTXO);
+}
+
+
+version (unittest)
+private PublicKey getTestKey (size_t idx)
+{
+    return KeyPair.fromSeed(Seed.fromString(pregen_seeds[idx])).address;
+}
+
+/// for debugging
+private string pretty (int[PublicKey] counts)
+{
+    string res;
+
+    int[string] result;
+    foreach (key, count; counts)
+        result[key_to_simple[key]] = count;
+
+    auto keys = result.keys;
+    sort!((a, b) => a[1..$].to!int < b[1..$].to!int)(keys);
+
+    foreach (key; keys)
+        res ~= format("%s: %s, ", key, result[key]);
+
+    return res;
+}
+
+// using pregenerated seeds to test the determenistic behavior of the algorithm
+// note: sorted for easier comparison with PublicKey[]
+version (unittest)
+private immutable string[64] pregen_seeds =
```

Comment:
This part made me realize something about how we use seeds in the test-suite. I'll file an issue for it.

Reply:
I've been wanting to add a module with "test data" for a while, I think it could be useful to have somewhere to get "valid, initialized data".

Reply:
Yeah. It would be nice to have consistent keys in our test-suite, it helps with debugging.